flask_dance/consumer/base.py

```diff
@@ -3,14 +3,21 @@
 import warnings
 from datetime import datetime, timedelta
 import six
-from lazy import lazy
 from abc import ABCMeta, abstractmethod, abstractproperty
 from werkzeug.datastructures import CallbackDict
 import flask
 from flask.signals import Namespace
 from flask_dance.consumer.storage.session import SessionStorage
 from flask_dance.utils import getattrd, timestamp_from_datetime
 
+try:
+    from werkzeug.utils import invalidate_cached_property
+except ImportError:
+    from werkzeug._internal import _missing
+
+    def invalidate_cached_property(obj, name):
+        obj.__dict__[name] = _missing
+
```

Comment:
I don't quite understand this block, what is this magical `_missing` thing?

Reply:
It's an implementation detail of how the `cached_property` decorator works. [If you check Werkzeug's source code](https://github.com/pallets/werkzeug/blob/master/src/werkzeug/utils.py#L89), you can see that `_missing` is a sentinel value that is returned if there is no value in `obj.__dict__[name]`, which works as a cache. In order to invalidate the cache, you have to reset that value to `_missing`, so that `cached_property` thinks that the cache is empty.

We should probably wait to merge this pull request until https://github.com/pallets/werkzeug/pull/1474 is merged -- at this point, the PR is still open, so the API could change.

Reply:
Thanks, that explains a few things!